plugin/pkg/admission/nodedeclaredfeatures/admission.go

```diff
@@ -0,0 +1,192 @@
+	nodeDeclaredFeatureFramework   *ndf.Framework
+	nodeDeclaredFeatureGateEnabled bool
+	version                        *versionutil.Version
+}
+
+var _ admission.ValidationInterface = &Plugin{}
+var _ = genericadmissioninitializer.WantsExternalKubeInformerFactory(&Plugin{})
+var _ genericadmissioninitializer.WantsFeatures = &Plugin{}
+
+// New creates a new Plugin admission plugin
+func NewPlugin() (*Plugin, error) {
+	framework, err := ndf.New(ndffeatures.AllFeatures)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create node declared features helper: %w", err)
+	}
+
+	ver, err := versionutil.Parse(version.Get().String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse version: %w", err)
+	}
+	return &Plugin{
+		Handler:                      admission.NewHandler(admission.Create, admission.Update),
+		nodeDeclaredFeatureFramework: framework,
+		version:                      ver,
+	}, nil
+}
+
+// SetExternalKubeInformerFactory sets the informer factory for the plugin.
+func (p *Plugin) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
+	nodeInformer := f.Core().V1().Nodes()
+	p.nodeLister = f.Core().V1().Nodes().Lister()
+	p.SetReadyFunc(nodeInformer.Informer().HasSynced)
+}
+
+// SetFeatures sets the feature gates for the plugin.
+func (p *Plugin) InspectFeatureGates(featureGates featuregate.FeatureGate) {
+	p.nodeDeclaredFeatureGateEnabled = featureGates.Enabled(features.NodeDeclaredFeatures)
+}
+
+// ValidateInitialization ensures that the plugin is properly initialized.
+func (p *Plugin) ValidateInitialization() error {
+	if p.nodeLister == nil {
+		return fmt.Errorf("missing node lister for %s", PluginName)
+	}
+
+	if p.nodeDeclaredFeatureFramework == nil {
+		framework, err := ndf.New(ndffeatures.AllFeatures)
+		if err != nil {
+			return fmt.Errorf("failed to create node feature helper for %s: %w", PluginName, err)
+		}
+		p.nodeDeclaredFeatureFramework = framework
+	}
+	return nil
+}
+
+// Validate is the core of the admission controller logic.
+func (p *Plugin) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	// Ignore if the feature gate is not enabled.
+	if !p.nodeDeclaredFeatureGateEnabled {
+		return nil
+	}
+	// We only care about Pod updates.
+	if a.GetResource().GroupResource() != core.Resource("pods") {
+		return nil
+	}
+
+	if a.GetOperation() != admission.Update {
+		return nil
+	}
+	// Only validate updates to the main pod spec (subresource == "")
+	// or the custom "resize" subresource.
+	subresource := a.GetSubresource()
+	if subresource != "" && subresource != "resize" {
+		return nil
+	}
+	pod, ok := a.GetObject().(*core.Pod)
+	if !ok {
+		return errors.NewBadRequest(fmt.Sprintf("expected a pod but got a %T", a.GetObject()))
+	}
+	oldPod, ok := a.GetOldObject().(*core.Pod)
+	if !ok {
+		return errors.NewBadRequest(fmt.Sprintf("expected an old pod but got a %T", a.GetOldObject()))
+	}
+
```

Comment:
I think we only care about main resource updates if the generation got bumped because the spec actually changed, right? can we do this as well?

```
if subresource == "" && pod.Generation == oldPod.Generation {
  return nil
}
```